src/seasidecontactbuilder.cpp

```diff
@@ -0,0 +1,666 @@
+    QVersitContactImporter importer;
+    importer.setPropertyHandler(handler);
+    importer.importDocuments(documents);
+    return importer.contacts();
+}
+
+/*
+ * Preprocess the given import contact prior to duplicate detection,
+ * merging, and subsequent storage.
+ *
+ * The default implementation performs some fixes for common issues
+ * encountered in NAME field details, and removes various detail types
+ * which are not supported by the qtcontacts-sqlite manager engine.
+ */
+void SeasideContactBuilder::preprocessContact(QContact &contact)
+{
+    // Fix up name (field ordering) if required
+    QContactName nameDetail = contact.detail<QContactName>();
+    if (applyNameFixes(&nameDetail)) {
+        contact.saveDetail(&nameDetail);
+    }
+
+    // Remove any details that our backend can't store, or which
+    // the client wishes stripped from the imported contacts.
+    foreach (QContactDetail detail, contact.details()) {
+        if (d->unimportableDetailTypes.contains(detail.type())) {
+            qDebug() << "  Removing unimportable detail:" << detail;
+            contact.removeDetail(&detail);
+        } else if (detail.type() == QContactSyncTarget::Type) {
+            // We allow some syncTarget values
+            const QString syncTarget(detail.value<QString>(QContactSyncTarget::FieldSyncTarget));
+            if (!d->importableSyncTargets.contains(syncTarget)) {
+                qDebug() << "  Removing unimportable syncTarget:" << syncTarget;
+                contact.removeDetail(&detail);
+            }
+        }
+    }
+
+    // Set nickname by default if the name is empty
+    if (contactNameString(contact).isEmpty()) {
+        QContactName nameDetail = contact.detail<QContactName>();
+        contact.removeDetail(&nameDetail);
+        if (contact.details<QContactNickname>().isEmpty()) {
+            QString label = contact.detail<QContactDisplayLabel>().label();
+            if (label.isEmpty()) {
+                label = SeasideCache::generateDisplayLabelFromNonNameDetails(contact);
+            }
+            setNickname(contact, label);
+        }
+    }
+}
+
+/*
+ * Returns the index into the \a importedContacts list at which a
+ * duplicate (merge candidate) of the \a contact at the current
+ * \a contactIndex may be found, or \c -1 if no match is found.
+ *
+ * The default implementation uses a combination of GUID and
+ * name / label matching to determine if a contact is duplicated
+ * within the import list.
+ *
+ * The result will be used to merge any duplicated contacts within
+ * the import list, which is the required behaviour when performing
+ * an "import" style sync.  Any implementation which requires
+ * a one-to-one mapping between import contacts and stored device
+ * contacts should instead return -1 from this function.
+ */
+int SeasideContactBuilder::previousDuplicateIndex(QList<QContact> &importedContacts, QContact &contact, int contactIndex)
```

Comment:
Is \a contact required to be at \a contactIndex?  If so, drop the \a contact parameter, it's just confusing.